main.cpp

```diff
@@ -73,418 +72,232 @@ compile/link/run to make sure you don't have any errors
 
-        FloatType& pow( const FloatType& operand );
-        FloatType& pow( const DoubleType& operand );
-        FloatType& pow( const IntType& operand );
-        FloatType& pow( float operand );
+    Numeric& operator-=( const Type y )
+    {
+        *heapNumber -= y;
+        return *this;
+    }
 
-        FloatType& apply(std::function<FloatType&(float&)> callable)
+    Numeric& operator*=( const Type y )
+    {
+        *heapNumber *= y;
+        return *this;
+    }
+  
+    Numeric& operator/=( const Type y )
+    {
+        if constexpr ( std::is_same<Type, int>::value )
         {
-            if(callable)
+            // self note: int is not the same as const int and your comparison must reflect that!
+            if constexpr ( std::is_same<decltype(y), const int>::value )
             {
-                return callable(*heapFloat); 
+                if ( y == 0 ) 
+                {
+                    std::cout << "Can't divide by integer 0! Cancelling operation." << std::endl;
+                    return *this; 
+                }
             }
-
-            return *this;
-        }
-        FloatType& apply( void(*funcPtr)(float&))
-        {
-            if(funcPtr)
+            else if ( y < std::numeric_limits<Type>::epsilon() )
             {
-                funcPtr(*heapFloat);
+                std::cout << "Can't divide by non-int 0! Cancelling operation." << std::endl;
+                return *this;
             }
-
-            return *this;
+                
         }
-
-    private:
-        float* heapFloat = nullptr; 
-        FloatType& powInternal( float x );
-
-};
-
-// double UDT
-struct DoubleType
-{
-    public:
-        DoubleType( double d ) : heapDub( new double(d) ) {}
-        ~DoubleType()
+        else if ( y < std::numeric_limits<Type>::epsilon() )
         {
-            delete heapDub;
-            heapDub = nullptr;
+            std::cout << "Warning! Dividing by 0." << std::endl;
         }
+            
+        *heapNumber /= y;
+        return *this;
+    }
 
-        operator double() const { return *heapDub; }
 
-        DoubleType& operator+=( const double y )
-        {
-            *heapDub += y;
-            return *this;
-        }
+    Numeric& pow( const Numeric& operand )
+    {
+        return powInternal( operand );
+    }
 
-        DoubleType& operator-=( const double y )
+    Numeric& apply(std::function<Numeric&(Type&)> callable)
+    {
+        if(callable)
         {
-            *heapDub -= y;
-            return *this;
+            return callable(*heapNumber); 
         }
 
-        DoubleType& operator*=( const double y )
-        {
-            *heapDub *= y;
-            return *this;
-        }
+        return *this;
+    }
 
-        DoubleType& operator/=( const double y )
+    Numeric& apply( void(*funcPtr)(Type&) )
+    {
+        if(funcPtr)
         {
-            *heapDub /= y;
-            return *this;
+            funcPtr(*heapNumber); 
         }
```

Comment:
Your `apply` function should take your unique_ptr, not the unique_ptr's held object
see instruction 4)
- [ ] revise